version_updates.md

```diff
@@ -0,0 +1,120 @@
+CR to apply a minor updates to a deployed OpenStack
+installation (including both Controlplane and Dataplane
+Services)
+
+## Requirements
+
+ * Use of OLM (Operator Lifecycle Manager) is assumed
+ * The CSV (ClusterServiceVersion) for the openstack-operator
+   is used as the primary resource for versioning and the
+   associated ENV variables for OpenStack service containers.
+   These are typically prefixed with RELATED\_IMAGE...
+   The CSV defines the version.
+ * OpenStack operators are always upgraded first via OLM. This
+   can be either via manual or automatic. Upgrading operators
+   will trigger OpenStackVersion to provide access to a new
+   available version.
+ * And ENV called OPENSTACK\_VERSION must be set in the CSV
+   for openstack-operator.  This controls the 'availableVersion'
+   of OpenStackVersion.
+   
+
+## General Update Workflow
+
+OpenStack is installed by following the normal installation procedure. When
+an OpenStackControlplane resource is created in any given namespace an
+OpenStackVersion resource will be immediately created when the OpenStackControlPlane
+is reconciled (1 OpenStackVersion per namespace).
+
+When a new set of Operators is installed via OLM (either manually or automatically)
+the OpenStackVersion resources will reconcile and a new 'AvailableVersion' will
+be set on each CR.
+
+In order to apply an update an administrator sets the 'TargetVersion' to the
+same value as the 'AvailableVersion' on the OpenStackVersion resources. This
+triggers the OpenStackVersion to reconcile and will update the images
+on the associated OpenStackControlplane and OpenStackDataplane resources in
+that namespace.
+
+```shell
+$ oc get openstackversion
+NAME              TARGET VERSION   AVAILABLE VERSION
+openstack         1.0.0            1.0.1
+```
+
+```yaml
+apiVersion: core.openstack.org/v1beta1
+kind: OpenStackVersion
+metadata:
+  labels:
+  name: openstack
+spec:
+  openstackControlPlaneName: openstack
+  targetVersion: 1.0.1
+```
+
+## Custom images for Cinder Volume
+
+In order to allow administrators to configure customized Cinder volume backends
+with custom containerImages for each backend you must exclude those images
+from automatic updates. This is because there is no way to fully automate
+the creation of those images at this time. To exclude Cinder volume named
+'netapp' from automatic updates you can add the name 'netapp' to the
+'cinderVolumeExcludes' array on the OpenStackVersion CR. Multiple named
+Cinder volume backends may be excluded.
+
+When the custom built containerImage(s) for a Cinder Volume have been
+created you can deploy them by editing/patching/updating the
+OpenStackControlplane directly.
+
+
+```yaml
+apiVersion: core.openstack.org/v1beta1
+kind: OpenStackVersion
+metadata:
+  labels:
+  name: openstack
+spec:
+  openstackControlPlaneName: openstack
+  targetVersion: 1.0.1
+  cinderVolumeExcludes:
+    -netapp
+```
+
+## Custom Images for Neutron
```

Comment:
I understand that you are targeting updates process specifically; but this new CR may be helpful in third party neutron ml2 driver integration. Let me know if this is out of scope.

---

To integrate a custom neutron ml2 driver, one would build a custom image with the driver python package (and any other dependencies). Then this custom image should replace the default one. How do we do it?  Are we allowed to override RELATED_IMAGE_..._DEFAULT for a particular deployment? Should there be serviceOverrides in addition to serviceExcludes?

Reply:
Need to update the docs today to reflect recent feedback. But custom ML2 drivers would get handled the same way we do custom Cinder volume backends. The administrator would edit/patch/update the OpenStackVersion spec so that a custom ML2 driver containerImage would get deployed instead of the default.

While an administrator could override RELATED_IMAGE_.. a change like that would be non-persistent on upgrades to the operator. Also, it would involve editing RELATED_IMAGE... values in multiple places (for different operator deployments). I think keeping the customizedImages in a CR might be cleaner.

Reply:
For Cinder we currently set the `containerImage` manually in the `OpenStackControlPlane` for the specific cinder volume backend: https://github.com/openstack-k8s-operators/cinder-operator/blob/5e8e748d6046beee57bf3e4271565ff3bb7f8d11/config/samples/backends/hpe/3par/iscsi/backend.yaml#L24

Reply:
@booxter I added expanded the section in this doc about Neutron. Still a TBD I think. There is currently no formal interface to specify custom ML2 drivers other than OVN (correct me if I am wrong here). So I think before we can add the ML2 section to this CR we should first define where and how those images would be used. As far as I know every other image we have listed on OpenStackVersion satisfies these conditions.